aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUser.java

```diff
@@ -767,6 +772,26 @@ public void getSession(final AuthenticationHandler callback) {
      */
     public Runnable initiateUserAuthentication(final AuthenticationDetails authenticationDetails,
             final AuthenticationHandler callback, final boolean runInBackground) {
+        final Runnable task = _initiateUserAuthentication(authenticationDetails, callback, runInBackground);
+        if (runInBackground) {
+            return new Runnable() {
+                @Override
+                public void run() {
+                    new Thread(new Runnable() {
```

Comment:
It's very hard to understand what is going on here. Do you have any test cases that cover these changes?

E.g. the changes in this commit makes `getSessionInBackground` call  `AuthenticationHandler#onFailure()` on _background_ thread once user enters (incorrect) username/password and `continueTask()` has been called. `onSuccess()` is still called from _main thread_ however. I would have expected the callbacks to run on main thread, or at least be consistent.

Reply:
There is a code fix that will be addressing the issue you have pointed out, the release is pending.

Reply:
Generally, this client calls the onSuccess and onFailure cases on the same thread that `foo()` is called. Except when the `fooInBackground()` variant is used, that is when the callbacks are made on the main thread's looper.